Sanitise HTML in product description [read-only]

anansilva opened the pull request:
:information_source: _Please use project **Discover Regenerative (Macdoch pt 2): #3A. Tech - OFN & OFN/DFC Endpoints** to track work on this issue._

#### What? Why?

- Part of #12448

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->

Sanitises the HTML of product description before reaching the database.

**Implementation comments/suggestion:**

Using our custom  `HtmlSanitizer` on `Spree::Product#description` breaks the Action Text editor. That is because Action Text includes tags that we are not allowed in our custom sanitiser. Action Text uses rails' default  `Rails::Html::SafeListSanitizer`.

I might have missed discussions regarding the tags and attributes that we want to allow at OFN but why not use Rails defaults?

For a complete list of the default allowed attributes/tags:
`Rails::Html::WhiteListSanitizer.allowed_attributes`
`Rails::Html::WhiteListSanitizer.allowed_tags`

Tests still pass since the defaults do not allow `<script>` attributes. And this way we don't break the Action Text editor, offering more text editing possibilities.

#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

1. [Editor] Visit Admin, Products, edit a product description. Make sure you can save the text with editing options provided by the editor.

<img width="953" alt="Screenshot 2024-05-31 at 16 43 35" src="https://github.com/openfoodfoundation/openfoodnetwork/assets/27692541/d41d4b92-8789-4e14-b208-4003a67e0744">


2. [Store] Visit Store, Select a hub and find the product that was edited in the first step. The description of the product should show up as it was edited in step 1.

<img width="1418" alt="Screenshot 2024-05-31 at 16 43 44" src="https://github.com/openfoodfoundation/openfoodnetwork/assets/27692541/19681962-e53a-4514-b0f8-ac4968590bd8">

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [ ] Feature toggled

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
<!-- Does this PR depend on another one?
     Add the link or remove this section. -->



#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
     List them here or remove this section. -->

rioug commented on app/models/spree/product.rb:
```diff
@@ -304,6 +304,16 @@ def variant_unit_with_scale=(variant_unit_with_scale)
 
+    # Remove any unsupported HTML.
+    def description
+      Rails::HTML::SafeListSanitizer.new.sanitize(super)
```
I am wondering if we should add this to our `HtmlSanitizer` class ? ie something like : `HtmlSanitizer.rails_sanitize` 
The idea is to have all the sanitizer in the same place so we can pick and choose the appropriate one, instead of calling various different sanitizer in various places.
@openfoodfoundation/core-devs, any though ?

dacook replied:
Agreed. In fact, if we can assume SafeListSanitizer is safe, then I  don't see why we have to specify an allowlist at all. 

Perhaps Maikel took the opinion of allowing only what is needed. In that case, I think we should keep one sanitising method, and add any elements necessary for ActionText to our `HtmlSanitizer`.

@mkllnk can you confirm your thoughts?

mkllnk replied:
Please proof me wrong but I thought that the default list can still be dangerous. For example, I think that it allows the `class` attribute. Some CSS classes are used by Javascript to add logic to elements. So suddenly someone could create a link that looks like a button in our design. It could be called _Checkout_ and lead to a fake payment site.

Does the editor only rely on HTML comments? We could probably allow those.

dacook replied:
Good point, we should add a case for that to `html_sanitizer_spec`, then we can prove it, right or wrong 😉

anansilva replied:
Great discussion! I've updated our custom sanitizer with the tags that are currently supported by our trix editor, and a specific `data-trix` attribute that is needed for the horizontal rule.

I've also added a more complete set of dangerous scenarios to the tests. This serves both as documentation of potential risks and as a proactive warning system through test failures should any new dangerous tags be introduced in the future.

dacook reviewed (changes requested):
Would you be able to try adding a spec for the case described? 
Then please try the default safe list. 
Then can you try making product description use our `HtmlSanitizer` class like the others? If using the allowlist, you can add any necessary tags to the list (eg html comments)

dacook reviewed (changes requested):
Great additions, thank you! 

There's a spec failure on this expectation:
https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/system/admin/products_spec.rb#L151

I would like to say let's get rid of the `div`, but I had a look and found a couple of style rules related to it. Can you please permit divs, or investigate to see if it can be removed?
* https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/system/admin/products_spec.rb#L151
* https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/webpacker/css/darkswarm/_shop-modals.scss#L15-L17

anansilva commented:
@dacook, trix does use `div` to wrap blocks so I've added it to the allowed tags. Also added a couple more tests to make sure we are warned if some tag is removed.

dacook commented on app/services/html_sanitizer.rb:
```diff
@@ -6,10 +6,15 @@
 # We offer an editor which supports certain tags but you can't insert just any
 # HTML, which would be dangerous.
 class HtmlSanitizer
+  ALLOWED_TAGS = %w[h1 h2 h3 h4 div p br b i u a strong em del pre blockquote ul ol li hr
```
It's nice and clear what's allowed now 👍 
Let's also add a comment to explain why:

```suggestion
  # div is required by Trix editor
  ALLOWED_TAGS = %w[h1 h2 h3 h4 div p br b i u a strong em del pre blockquote ul ol li hr
```

dacook reviewed (approved):
Great job, thank you

mkllnk reviewed (approved):
Excellent! :star2:

filipefurtad0 commented:
Hey @anansilva ,

I've staged the PR and verified that:
- editing a product description and saving changes is reflected in the shopfront.
- I've tested several options available by the editor

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/c9848e5b-492a-45c5-a6fa-18fbd1581470)

Looking good :-)
Merging.
